cpp/oneapi/dal/algo/knn/detail/model_impl.hpp

```diff
@@ -0,0 +1,31 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/knn/common.hpp"
+
+namespace oneapi::dal::knn {
+
+class detail::model_impl: public base {
```

Comment:
```c++
// file: dal/detail/common.hpp
class pimpl_accessor {
    template <typename Object, typename ... Args>
    static auto make(Args&& ...args) {
        return Object{ std::forward<Args>(args)... };
    }
};


// file: common.hpp
namespace detail {
class model_impl;
} // namespace detail

class model {

};


// file: detail/model_impl.hpp
namespace detail {

class model_impl {
public:
    // If needed
    virtual foo() = 0;
};

} // namespace detail


// file: backend/interop_model.hpp
#include "detail/model_impl.hpp"
namespace backend {

class daal_interop_model : public model_impl {

};

} // namespace backend


// file: common.hpp
class model {
friend dal::detail::pimpl_accessor;
public:
    model();

private:
    explicit model(const std::shared_ptr<detail::model_impl>& impl);

    dal::detail::pimpl<detail::model_impl> impl_;
};


// file: common.cpp
#include "detail/model_impl.hpp"

class empty_model_impl : public detail::model_impl {};

// Default constructor
model::model() : impl_(new empty_model_impl{}) {}

model::model(const std::shared_ptr<detail::model_impl>& impl) : impl_(impl) {}


// file: backend/cpu/kernel.cpp
#include "backend/interop_model.hpp"

model train() {
    const auto model_impl = std::make_shared<daal_interop_model>(...);
    return dal::detail::pimpl_accessor::make<model>(model_impl);
}

```

Reply:
I like the idea of empty model. Will rework.